core/src/main/scala/aima/core/agent/Environment.scala

```diff
@@ -3,9 +3,7 @@ package aima.core.agent
 /**
   * @author Shawn Garner
   */
-trait Environment[Action, Percept] {
-  def addAgent(agent: Agent[Action, Percept]): Environment[Action, Percept]
-  def removeAgent(agent: Agent[Action, Percept]): Environment[Action, Percept]
-  def actuate(actuator: Actuator[Action, Percept], action: Action): Environment[Action, Percept]
-  def perceive(sensor: Sensor[Action, Percept]): Percept
+trait Environment[ENVIRONMENT, PERCEPT, ACTION] {
+  def addAgent(agent: Agent[ENVIRONMENT, PERCEPT, ACTION]): ENVIRONMENT
+  def removeAgent(agent: Agent[ENVIRONMENT, PERCEPT, ACTION]): ENVIRONMENT
```

Comment:
@benderpremier The reason I had actuate and perceive in the environment is that the environment kind of dictates the rules of what can be perceived and what can actions can be performed on it.  They agent may have certain sensors or actuators to do things but it can not fundamentally do something that is not permissible by the environment.  I'm thinking of games where I could as an actor in the environment just change the environment to teleport me to one spot or another or give me infinite resources.  It seems weird that the agent would be directly in control over manipulating the environment.